### PR TITLE
DEVOPS-452: Update package with python-conda-template

### PR DESCRIPTION
**Pull request by SophieCurinier**
**DEVOPS-452 - apply latest version of python template repos to all repos**
- Use reusable github workflows
- Use MiraGeoscience/pre-commit-hooks package in `.pre-commit-hooks.yml`
- Update pre-commit-hooks version:
   - `ruff-pre-commit` : v0.5.5
   - `mirrors-mypy`: v1.10.1
   - `Mirageoscience/pre-commit-hooks`: v1.0.1
   - `codespell`: v2.3.0
   - `rstcheck`: v6.2.4

### CODE REVIEW

File: simpeg_drivers/components/factories/survey_factory.py [hunk 1]
```diff
@@ -375,7 +375,7 @@ def _tdem_arguments(self, data=None, local_index=None, mesh=None):
         tx_list = []
         rx_factory = ReceiversFactory(self.params)
         tx_factory = SourcesFactory(self.params)
-        for tx_locs, rx_ids in zip(tx_locs, rx_lookup):
+        for tx_locs, rx_ids in zip(tx_locs, rx_lookup, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/components/factories/survey_factory.py:378]: **question:** strict=True? is it expected to always have len() == len()?

@domfournier  @benk-mira  ?

File: simpeg_drivers/components/meshes.py [hunk 2]
```diff
@@ -183,7 +184,7 @@ def ensure_cell_convention(mesh: Octree) -> TreeMesh | None:
             raise ValueError("Cannot convert negative cell sizes for rotated mesh.")
 
         cell_sizes, origin = [], []
-        for axis, dim in zip("xyz", "uvw"):
+        for axis, dim in zip("xyz", "uvw", strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/components/meshes.py:187]:
```suggestion
        for axis, dim in zip("xyz", "uvw", strict=True):
```

File: simpeg_drivers/joint/driver.py [hunk 3]
```diff
@@ -54,7 +54,7 @@ def data_misfit(self):
         if getattr(self, "_data_misfit", None) is None and self.drivers is not None:
             objective_functions = []
             multipliers = []
-            for label, driver in zip("abc", self.drivers):
+            for label, driver in zip("abc", self.drivers, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/joint/driver.py:57]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/joint/driver.py [hunk 4]
```diff
@@ -124,7 +124,7 @@ def initialize(self):
             global_actives |= local_actives
 
         self.models.active_cells = global_actives
-        for driver, wire in zip(self.drivers, self.wires):
+        for driver, wire in zip(self.drivers, self.wires, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/joint/driver.py:127]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/joint/driver.py [hunk 5]
```diff
@@ -205,7 +205,7 @@ def run(self):
                 self.models.starting, compute_J=False
             )
 
-            for sub, driver in zip(predicted, self.drivers):
+            for sub, driver in zip(predicted, self.drivers, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/joint/driver.py:208]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/joint/joint_cross_gradient/driver.py [hunk 6]
```diff
@@ -107,17 +103,14 @@ def directives(self):
...
+                    if driver_directives.vector_inversion_directive is not None:
                         directives_list.append(
-                            getattr(driver_directives, "vector_inversion_directive")
+                            driver_directives.vector_inversion_directive
                         )
 
                     count += n_tiles
 
-                for driver, wire in zip(self.drivers, self.wires):
+                for driver, wire in zip(self.drivers, self.wires, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/joint/joint_cross_gradient/driver.py:113]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/joint/joint_cross_gradient/driver.py [hunk 7]
```diff
@@ -170,7 +163,7 @@ def get_regularization(self):
             driver.regularization = ComboObjectiveFunction(objfcts=reg_block)
 
         for label, driver_pairs in zip(
-            ["a_b", "c_a", "c_b"], combinations(self.drivers, 2)
+            ["a_b", "c_a", "c_b"], combinations(self.drivers, 2), strict=False
```
Comment (sebhmg) [inline simpeg_drivers/joint/joint_cross_gradient/driver.py:166]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/utils/utils.py [hunk 8]
```diff
@@ -114,7 +115,7 @@ def calculate_2D_trend(
 
     polynomial = []
     xx, yy = np.triu_indices(order + 1)
-    for x, y in zip(xx, yy):
+    for x, y in zip(xx, yy, strict=False):
```
Comment (sebhmg) [inline simpeg_drivers/utils/utils.py:118]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/utils/utils.py [hunk 9]
```diff
@@ -129,7 +130,7 @@ def calculate_2D_trend(
 
     params, _, _, _ = np.linalg.lstsq(polynomial, values, rcond=None)
     data_trend = np.zeros(points.shape[0])
-    for count, (x, y) in enumerate(zip(xx, yy)):
+    for count, (x, y) in enumerate(zip(xx, yy, strict=False)):
```
Comment (sebhmg) [inline simpeg_drivers/utils/utils.py:133]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/utils/utils.py [hunk 10]
```diff
@@ -461,7 +460,7 @@ def get_inversion_output(h5file: str | Workspace, inversion_group: str | UUID):
     ][:-1]
     cols = out.pop(0).split(" ")
     out = [[string_to_numeric(k) for k in elem.split(" ")] for elem in out]
-    out = dict(zip(cols, list(map(list, zip(*out)))))
+    out = dict(zip(cols, list(map(list, zip(*out, strict=False))), strict=False))
```
Comment (sebhmg) [inline simpeg_drivers/utils/utils.py:463]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: tests/run_tests/driver_fem_test.py [hunk 11]
```diff
@@ -118,7 +119,7 @@ def test_fem_run(tmp_path: Path, max_iterations=1, pytest=True):
 
         data_kwargs = {}
         for comp, data_group, uncert_group in zip(
-            components, data_groups, uncert_groups
+            components, data_groups, uncert_groups, strict=False
```
Comment (sebhmg) [inline tests/run_tests/driver_fem_test.py:122]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: tests/run_tests/driver_mt_test.py [hunk 12]
```diff
@@ -135,7 +136,7 @@ def test_magnetotellurics_run(tmp_path: Path, max_iterations=1, pytest=True):
 
         data_kwargs = {}
         for comp, data_group, uncert_group in zip(
-            components, data_groups, uncert_groups
+            components, data_groups, uncert_groups, strict=False
```
Comment (sebhmg) [inline tests/run_tests/driver_mt_test.py:139]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: tests/run_tests/driver_tipper_test.py [hunk 13]
```diff
@@ -113,7 +114,7 @@ def test_tipper_run(tmp_path: Path, max_iterations=1, pytest=True):
 
         data_kwargs = {}
         for comp, data_group, uncert_group in zip(
-            components, data_groups, uncert_groups
+            components, data_groups, uncert_groups, strict=False
```
Comment (sebhmg) [inline tests/run_tests/driver_tipper_test.py:117]: **question:** use `strict = True`?

@domfournier or @benk-mira  ?

File: simpeg_drivers/joint/driver.py (hunk 3 above)
Comment (domfournier) [reply simpeg_drivers/joint/driver.py:57]: Yes, let's do true.

File: simpeg_drivers/utils/utils.py (hunk 9 above)
Comment (domfournier) [reply simpeg_drivers/utils/utils.py:133]: You're repeating yourself bra

**Review by sebhmg — APPROVED** [review]

**Comment by sebhmg** [comment]
are tests failing since the changes with `strict=True`?

**Comment by SophieCurinier** [comment]
Thank you!

**Pull request merged**